File: solutions/cpp/0024.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head || !head->next) return head;

        ListNode* dummy = new ListNode(NULL);
        dummy->next = head;

        int len = 0;
        for (auto curr = head; curr; curr = curr->next) len++;

        auto prev = dummy;
        auto curr = head;
        for (int i = 0; i < len / 2; i++) {
            auto next = curr->next;
            curr->next = next->next;
            next->next = prev->next;
            prev->next = next;
            prev = curr;
            curr = curr->next;
        }

        return dummy->next;
    }
};

File: solutions/python3/0024.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head

        dummy = ListNode(None)
        dummy.next = head

        len = 0
        curr = head
        while curr:
            len += 1
            curr = curr.next

        prev = dummy
        curr = head
        for i in range(len // 2):
            next = curr.next
            curr.next = next.next
            next.next = prev.next
            prev.next = next
            prev = curr
            curr = curr.next

        return dummy.next
